src/full/Agda/TypeChecking/Serialise/Base.hs

```diff
@@ -362,57 +449,62 @@ icodeMemo getDict getCounter a icodeP = do
 --   via the @valu@ function and stores it in 'nodeMemo'.
 --   If @ix@ is present in 'nodeMemo', @valu@ is not used, but
 --   the thing is read from 'nodeMemo' instead.
-vcase :: forall a . EmbPrj a => (Node -> R a) -> Int32 -> R a
+vcase :: forall a . EmbPrj a => ([Int32] -> R a) -> Int32 -> R a
 vcase valu = \ix -> do
     memo <- gets nodeMemo
-    -- compute run-time representation of type a
-    let aTyp = typeRep (Proxy :: Proxy a)
+    let fp = fingerprintNoinline (typeRep (Proxy :: Proxy a))
     -- to introduce sharing, see if we have seen a thing
     -- represented by ix before
     slot <- liftIO $ readArray memo ix
-    case Hm.lookup aTyp slot of
-      -- yes, we have seen it before, use the version from memo
-      Just (U u) -> maybe malformed return (cast u)
-      -- no, it's new, so generate it via valu and insert it into memo
-      Nothing    -> do
-          v <- valu . (! ix) =<< gets nodeE
-          liftIO $ writeArray memo ix (Hm.insert aTyp (U v) slot)
-          return v
+    lookupME (Proxy :: Proxy a) fp slot
+      -- use the stored value
+      pure
+      -- read new value and save it
+      do !v <- valu . (! ix) =<< gets nodeE
+         liftIO $ writeArray memo ix $! MECons fp (unsafeCoerce v) slot
+         return v
 
 -- | @icodeArgs proxy (a1, ..., an)@ maps @icode@ over @a1@, ..., @an@
 --   and returns the corresponding list of @Int32@.
 
 class ICODE t b where
   icodeArgs :: IsBase t ~ b => All EmbPrj (Domains t) =>
-               Proxy t -> Products (Domains t) -> S [Int32]
+               Proxy t -> StrictProducts (Domains t) -> S Node
 
 instance IsBase t ~ 'True => ICODE t 'True where
-  icodeArgs _ _  = return []
+  icodeArgs _ _  = return Empty
+  {-# INLINE icodeArgs #-}
 
 instance ICODE t (IsBase t) => ICODE (a -> t) 'False where
-  icodeArgs _ (a , as) = icode a >>= \ hd -> (hd :) <$> icodeArgs (Proxy :: Proxy t) as
+  icodeArgs _ (Pair a as) = do
+    !hd   <- icode a
+    !node <- icodeArgs (Proxy :: Proxy t) as
+    pure $ Cons hd node
```

Comment:
I am trying to understand whether a `$!` would make a difference here.
Below you have it, but no strict bindings of `hd` and `node`.
Are both ways equivalent?

Reply:
˙($!)` doesn't make a diff here because `Cons hd node` is a redex-free constructor expression for which GHC never creates a thunk. Alternatively, we can omit bangs from `hd` and `node`, and use `($!)`. That works because the fields of `Cons` are strict, so if `Cons` is whnf-ed then `hd` and `node` must be also. This would not work with lazy fields.